base/setup/reactos/drivepage.c

```diff
@@ -48,28 +48,254 @@ static const INT  column_alignment[MAX_LIST_COLUMNS] = {LVCFMT_LEFT, LVCFMT_LEFT
+            return TRUE;
+        __fallthrough;
+
+    case WM_IME_CHAR:
+    case WM_CHAR:
+    {
+        WCHAR wch = (WCHAR)wParam;
+
+        /* Let the EDIT control deal with Control characters.
+         * It won't emit them as raw data in the text. */
+        if (wParam < ' ') // if (GetKeyState(VK_CONTROL) & 0x8000)
+            break;
+
+        /* Ignore Ctrl-Backspace */
+        if (wParam == '\x7F')
+            return 0;
+
+        /* Protect read-only edit control from modification */
+        if (GetWindowLongPtrW(hWnd, GWL_STYLE) & ES_READONLY)
+            break;
+
+        if (uMsg == WM_IME_CHAR)
+        {
+            if (!IsWindowUnicode(hWnd) && HIBYTE(wch) != 0)
+            {
+                CHAR data[] = {HIBYTE(wch), LOBYTE(wch)};
+                MultiByteToWideChar(CP_ACP, 0, data, 2, &wch, 1);
+            }
+        }
+
+        /* Show an error and ignore input character if it's invalid */
+        if (!IS_VALID_INSTALL_PATH_CHAR(wch))
+        {
+            ShowErrorTip(hWnd);
+            return 0;
+        }
+        break;
+    }
+
+    case WM_PASTE:
+        /* Verify the text being pasted; if it was sanitized, show an error */
+        if (DoSanitizeClipboard(hWnd))
+            ShowErrorTip(hWnd);
+        break;
+    }
+
+    return CallWindowProcW(orgEditProc, hWnd, uMsg, wParam, lParam);
+}
+
+static INT_PTR
+CALLBACK
+MoreOptDlgProc(
+    _In_ HWND hDlg,
+    _In_ UINT uMsg,
+    _In_ WPARAM wParam,
+    _In_ LPARAM lParam)
 {
     PSETUPDATA pSetupData;
 
     /* Retrieve pointer to the global setup data */
-    pSetupData = (PSETUPDATA)GetWindowLongPtrW(hwndDlg, GWLP_USERDATA);
+    pSetupData = (PSETUPDATA)GetWindowLongPtrW(hDlg, GWLP_USERDATA);
 
     switch (uMsg)
     {
         case WM_INITDIALOG:
         {
+            HWND hEdit;
+            WNDPROC orgEditProc;
+            BOOL bIsBIOS;
+            UINT uID;
+            INT iItem, iCurrent = CB_ERR, iDefault = 0;
+            WCHAR szText[50];
+
             /* Save pointer to the global setup data */
             pSetupData = (PSETUPDATA)lParam;
-            SetWindowLongPtrW(hwndDlg, GWLP_USERDATA, (DWORD_PTR)pSetupData);
+            SetWindowLongPtrW(hDlg, GWLP_USERDATA, (LONG_PTR)pSetupData);
+
+            /* Subclass the install-dir edit control */
+            hEdit = GetDlgItem(hDlg, IDC_PATH);
+            orgEditProc = (WNDPROC)GetWindowLongPtrW(hEdit, GWLP_WNDPROC);
+            SetWindowLongPtrW(hEdit, GWLP_USERDATA, (LONG_PTR)orgEditProc);
+            SetWindowLongPtrW(hEdit, GWLP_WNDPROC, (LONG_PTR)InstallDirEditProc);
+
+            /* Set the current installation directory */
+            SetWindowTextW(hEdit, pSetupData->USetupData.InstallationDirectory);
+
+
+            /* Initialize the list of possible bootloader locations */
+            bIsBIOS = ((pSetupData->USetupData.ArchType == ARCH_PcAT) ||
+                       (pSetupData->USetupData.ArchType == ARCH_NEC98x86));
```

Comment:
This can be kept as is, bootloader installation on Xbox is as simple as copying freeldr.sys onto the disk (Cromwell will boot it).

NEC PC-98 uses IPL boot scheme on hard disks (same for FM Towns), which is different from MBR, but for simplicity of terms we could still call it MBR.

Reply:
Indeed; well, the "MBR" sector on NEC (what's going to be executed by its firmware) is `<jump><IPL_signature><code><8_partition_entries_list><padding><0x55AA>`, as far as I remember, which is quite similar to BIOS-based PC-AT MBR. (The latter doesn't have the IPL signature, and only has 4 partition entries. But that doesn't matter in that part of the code.)
And that NEC MBR code could always start whatever is in a given partition's VBR, no? I think so...

Reply:
FMR/NEC MBR code starts with "IPLn" signature at offset 0, where `n` is related to the [device family](https://wiki3.jp/fmtowns/page/195). The firmware looks for that signature and then executes code directly from offset 4, and there's no `0x55AA` at the sector end.

Reply:
There's also a different NEC scheme, unrelated to PC-98: https://thestarman.pcministry.com/asm/mbr/DOS33MBR.htm

Reply:
Will check these; it seems we would have to write different "MBR" boot sectors for these in any case.

Reply:
> There's also a different NEC scheme, unrelated to PC-98: https://thestarman.pcministry.com/asm/mbr/DOS33MBR.htm

That one is quite similar to standard MBR, but with extra stuff to mark and support a 8-entry partition list.

Reply:
Just double checked, yes you were right, NEC PC-98 uses exactly this layout: `<jump><IPL_signature><code><8_partition_entries_list><padding><0x55AA>` - so the IPL1 signature is at offset 4 (contrary to Fujitsu FMR IPL bootsector).